### ISSUE TITLE
Why not run middleware on `undefined` instances?

### ISSUE THREAD

**Issue by rdallasgray** [young-steveo/bottlejs#43]
Hey -- see this line: https://github.com/young-steveo/bottlejs/blob/master/src/Bottle/provider.js#L89

We've recently hit an issue where a typo caused something not to be registered, and the error didn't show up until runtime. We'd like to use middleware to check for undefined instances being returned from the container, but as middleware isn't run on undefined instances, that won't work.

Is there a reason for not running middleware on undefined instances?

**Comment by young-steveo** [comment]
@rdallasgray Could you give me an example of a typo that would still execute the middleware?  As it stands now, a property getter is defined when a service is registered.  That getter will only fire if the property is accessed, so if the defining code or the calling code are mismatched, the getter will never fire, so the middleware will never fire.

Under the hood, this is approximately what bottle is doing:

```js
container = {};
Object.defineProperty(container, 'Dervice', { // note the typo
    configurable : true,
    enumerable : true,
    get : function() {
        // do the middleware stuff and return the instance
    }
});

container.Service; // undefined, because of the typo Dervice above
```

As a matter of fact, I removed the [undefined check in question](https://github.com/young-steveo/bottlejs/blob/master/src/Bottle/provider.js#L89) in a local build, and all of the unit tests pass.  So, I could probably remove the check, but it will likely not solve your problem.

**Comment by rdallasgray** [comment]
Ah, OK. That wouldn't work then.

What we're looking for is a fast and noisy failure when we try to access a service which hasn't been registered. The typo in question was comparable to your example above -- a misspelling of a service name which resulted in the 'real' service not being available at runtime, so we got an undefined passed to a collaborating object instead of the real service.

Now that I'm writing this of course I see that the real answer would be better error checking in the collaborating object ... but it may still be useful to have a way to fail fast when non-existent services are accessed. If you agree I can look at creating a PR?

**Comment by young-steveo** [comment]
I hadn't considered the implied injection of `bottle.service`.  I don't think it's possible to throw a semantic error when accessing a misspelled service from the container using a property lookup, but we should be able to do so if the automatic injection fails.  i.e.:

```js

 // this will be undefined, and there's not much we can do about that.
bottle.container.SpelledWrong;

// This could throw a nice error message for the user, since we're in control.
bottle.service('Parent', Parent, 'ChildA', 'SpelledWrong');
bottle.container.Parent;
```

**Comment by young-steveo** [comment]
I've got a PR coming today that should throw an exception if `service` or `resolve` encounter an undefined value.

**Comment by rdallasgray** [comment]
Awesome, thanks so much.

We now have three apps in production using Bottle, and I can honestly say it's been the answer to a lot of questions for us in making large JS apps. Thanks!

**Comment by young-steveo** [comment]
Hey you're welcome, thanks!

**Comment by young-steveo** [comment]
Released in [v1.2.2](https://github.com/young-steveo/bottlejs/releases/tag/v1.2.2)
